pkg/networkservice/common/connect/server.go

```diff
@@ -36,6 +39,19 @@ type connectServer struct {
 	clientFactory           func(ctx context.Context, conn grpc.ClientConnInterface) networkservice.NetworkServiceClient
 	uRLToClientMap          clientMap // key == url as string
 	connectionIDToClientMap clientMap // key == connection.GetId()
+	dialOptionFactories     []func(ctx context.Context, request *networkservice.NetworkServiceRequest, clientURL *url.URL) []grpc.DialOption
+}
+
+type dialOptionFactory struct {
+	factory func(ctx context.Context, request *networkservice.NetworkServiceRequest, clientURL *url.URL) []grpc.DialOption
+	grpc.EmptyDialOption
+}
+
+// WithDialOptionFactory - define a dial option factory, will be called on moment to create dial
```

Comment:
Could you give an example of where we need a DialOptionFactory rather than just a DialOption?

Reply:
https://github.com/networkservicemesh/cmd-nsmgr/pull/8/files#diff-52369f926d2c6c933dce1bfafc6cb6efR79

Reply:
Factory is required, since we do not know a endpoint uri until Request will be processed, so for callbacks we need to modify dial options.